include/seastar/coroutine/async_generator.hh

```diff
@@ -0,0 +1,421 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include <cassert>
+#include <coroutine>
+#include <exception>
+#include <iterator>
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <seastar/core/future.hh>
+
+// async_generator concept is heavily inspired by P2502R2
+// (https://wg21.link/P2502R2), a proposal accepted into C++23. P2502R2
+// introduced std::generator, which provides a synchronous coroutine
+// mechanism for generating ranges. in contrast, async_generator offers
+// asynchronous generation of element sequences.
+namespace seastar::coroutine::experimental {
+
+template<typename Ref, typename Value = void>
+class async_generator;
+
+namespace internal {
+
+template <typename Yielded> class next_awaiter;
+
+template <typename Yielded>
```

Comment:
Should have a constraint (type suitable to be carried in a future).

Reply:
in this implementation, the return value is passed to the consumer via a pointer -- the producer assigns the value to the pointer, so `seastar::future<>` is not involved.